File: VB/TreeListFilter/FilterTreeList/ColumnFilter/ColumnFilterEngine.vb
Imports System
Imports System.Collections.Generic
Imports System.Drawing
Imports System.Windows.Forms
Imports DevExpress.XtraTreeList
Imports DevExpress.XtraTreeList.Columns

Namespace FilterTreeListControl
	Public Class ColumnFilterEngine
		Implements IDisposable
		Private _ownerTreeList As FilterTreeList
		Private filterPopupForm As PopupForm
		Private _filterShowingColumn As TreeListColumn

		Public Sub New(ByVal owner As FilterTreeList)
			_ownerTreeList = owner
		End Sub

		Protected Overridable Function CreatePopupForm() As PopupForm
			Return New PopupForm(_ownerTreeList)
		End Function

		Private Sub ShowPopupForm(ByVal filterButtonBounds As Rectangle)
			Dim filterPopupFormBounds As Rectangle = filterPopupForm.CalcPopupRect(filterButtonBounds)
			filterPopupForm.Bounds = filterPopupFormBounds
			filterPopupForm.Show()
		End Sub

		Public Sub ProcessFilterButtonClick(ByVal column As TreeListColumn, ByVal filterButtonBounds As Rectangle)
			_filterShowingColumn = column

			filterPopupForm = CreatePopupForm()
			AddHandler filterPopupForm.FormClosing, AddressOf filterPopupFormOnFormClosing
			PopulateFilterList()
			ShowPopupForm(filterButtonBounds)
		End Sub

		Protected Overridable Function GetFilterValues() As List(Of ColumnFilterItem)
			Dim nodesOperation As New NodesValuesOperation(_filterShowingColumn)
			OwnerTreeList.NodesIterator.DoOperation(nodesOperation)
			Return nodesOperation.Result
		End Function

		Private Shared Sub RemoveColumnDataDuplicates(ByVal columnValuesList As List(Of ColumnFilterItem))
			Dim item As ColumnFilterItem = columnValuesList(0)
			Dim sameIndex As Integer = 0
			Dim sameCount As Integer = 0
			Dim i As Integer = 1

			Do While i < columnValuesList.Count
				If columnValuesList(i) = item Then
					sameCount += 1
					i += 1
				Else
					If sameCount > 0 Then
						columnValuesList.RemoveRange(sameIndex, sameCount)
						i = i - sameCount
						sameCount = 0
						sameIndex = i
						item = columnValuesList(i)
						i += 1
					Else
						sameIndex = i
						item = columnValuesList(i)
						i += 1
					End If
				End If
			Loop

			If sameCount > 0 Then
				columnValuesList.RemoveRange(sameIndex, sameCount)
			End If
		End Sub

		Private Shared Function IsFilterItemValueEmpty(ByVal value As ColumnFilterItem) As Boolean
			Return value.IsEmpty
		End Function

		Protected Sub PopulateFilterList()
			Dim columnValuesList As List(Of ColumnFilterItem) = GetFilterValues()
			If columnValuesList.Count = 0 Then
				Return
			End If

			columnValuesList.Sort(New FilterItemsComparer())
			columnValuesList.Insert(0, New ColumnFilterCustomItem())
			Dim blanksItemIndex As Integer = 1
			If OwnerTreeList.ColumnFilterConditions.ContainsColumn(_filterShowingColumn) Then
				columnValuesList.Insert(0, New ColumnFilterAllItem())
				blanksItemIndex += 1
			End If

			RemoveColumnDataDuplicates(columnValuesList)
			If columnValuesList.RemoveAll(AddressOf IsFilterItemValueEmpty) <> 0 Then
				columnValuesList.Insert(blanksItemIndex, New ColumnFilterBlanksItem())
				columnValuesList.Insert(blanksItemIndex + 1, New ColumnFilterNonBlanksItem())
			End If

			filterPopupForm.filterItemsList.Items.Clear()
			filterPopupForm.filterItemsList.Items.AddRange(columnValuesList.ToArray())
		End Sub

		Private Sub AddSingleFilterCondition(ByVal filterConditionString As String, ByVal filterValue As ColumnFilterItem)
			OwnerTreeList.FilteredVisibleNodes.Clear()

			Dim condition As FilterConditionEnum
			Dim newCondition As ColumnFilterCondition

			If TypeOf filterValue Is ColumnFilterBlanksItem Then
				newCondition = New ColumnFilterCondition(_filterShowingColumn.FieldName, FilterConditionEnum.IsBlank, Nothing, Nothing)
			ElseIf TypeOf filterValue Is ColumnFilterNonBlanksItem Then
				newCondition = New ColumnFilterCondition(_filterShowingColumn.FieldName, FilterConditionEnum.IsNotBlank, Nothing, Nothing)
			Else
				condition = CType(System.Enum.Parse(GetType(FilterConditionEnum), filterConditionString), FilterConditionEnum)
				newCondition = New ColumnFilterCondition(_filterShowingColumn.FieldName, condition, filterValue.Value, filterValue.DisplayText)
			End If

			For i As Integer = 0 To OwnerTreeList.ColumnFilterConditions.Count - 1
				If OwnerTreeList.ColumnFilterConditions(i).Column Is _filterShowingColumn Then
					OwnerTreeList.ColumnFilterConditions.RemoveAt(i)
					Exit For
				End If
			Next i

			If Not(TypeOf filterValue Is ColumnFilterAllItem) Then
				OwnerTreeList.ColumnFilterConditions.Add(newCondition)
			End If
		End Sub

		Protected Overridable Function CreateCustomFilterForm() As CustomFilterForm
			Return New CustomFilterForm()
		End Function

		Protected Overridable Sub PopuplateFilterConditions(ByVal frm As CustomFilterForm)
			frm.filterGroupPanel.Text = _filterShowingColumn.GetCaption()
			For i As Integer = 0 To OwnerTreeList.ColumnFilterConditions.Count - 1
				If OwnerTreeList.ColumnFilterConditions(i).Column Is _filterShowingColumn Then
					frm.cbeFilterConditions.Text = OwnerTreeList.ColumnFilterConditions(i).Condition.ToString()
					frm.teValue.Text = OwnerTreeList.ColumnFilterConditions(i).Value.ToString()
					Exit For
				End If
			Next i
		End Sub

		Protected Overridable Function GetCustomFilterValue(ByVal frm As CustomFilterForm) As ColumnFilterItem
			Dim filterValue As ColumnFilterItem
			Try
				filterValue = New ColumnFilterItem(Convert.ChangeType(frm.teValue.EditValue, _filterShowingColumn.ColumnType), frm.teValue.Text)
			Catch
				filterValue = New ColumnFilterItem(frm.teValue.Text, frm.teValue.Text)
			End Try

			Return filterValue
		End Function

		Private Sub ProcessItemSelection(ByVal sender As Object)
			Dim filterValue As ColumnFilterItem = CType((CType(sender, PopupForm)).SelectedFilterItem, ColumnFilterItem)
			If filterValue Is Nothing Then
				Return
			End If

			If TypeOf filterValue Is ColumnFilterCustomItem AndAlso _filterShowingColumn IsNot Nothing Then
				Dim frm As CustomFilterForm = CreateCustomFilterForm()
				PopuplateFilterConditions(frm)

				If frm.ShowDialog() = DialogResult.OK Then
					filterValue = GetCustomFilterValue(frm)
					AddSingleFilterCondition(frm.cbeFilterConditions.Text, filterValue)
				End If
			ElseIf (Not filterValue.IsEmpty) AndAlso _filterShowingColumn IsNot Nothing Then
				AddSingleFilterCondition("Equals", filterValue)
			End If

			_filterShowingColumn = Nothing
		End Sub

		Private Sub filterPopupFormOnFormClosing(ByVal sender As Object, ByVal e As FormClosingEventArgs)
			ProcessItemSelection(sender)
		End Sub

		Public Sub Dispose() Implements IDisposable.Dispose
			_ownerTreeList = Nothing
			_filterShowingColumn = Nothing
		End Sub

		Public ReadOnly Property OwnerTreeList() As FilterTreeList
			Get
				Return _ownerTreeList
			End Get
		End Property

		Public ReadOnly Property FilterShowingColumn() As TreeListColumn
			Get
				Return _filterShowingColumn
			End Get
		End Property
	End Class
End Namespace

File: CS/TreeListFilter/FilterTreeList/ColumnFilter/ColumnFilterEngine.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Windows.Forms;
using DevExpress.XtraTreeList;
using DevExpress.XtraTreeList.Columns;

namespace FilterTreeListControl
{
	public class ColumnFilterEngine : IDisposable
	{
		private FilterTreeList ownerTreeList;
		private PopupForm filterPopupForm;
		private TreeListColumn filterShowingColumn;

		public ColumnFilterEngine(FilterTreeList owner)
		{
			ownerTreeList = owner;
		}

		protected virtual PopupForm CreatePopupForm()
		{
			return new PopupForm(ownerTreeList);
		}

		private void ShowPopupForm(Rectangle filterButtonBounds)
		{
			Rectangle filterPopupFormBounds = filterPopupForm.CalcPopupRect(filterButtonBounds);
			filterPopupForm.Bounds = filterPopupFormBounds;
			filterPopupForm.Show();
		}

		public void ProcessFilterButtonClick(TreeListColumn column, Rectangle filterButtonBounds)
		{
			filterShowingColumn = column;

			filterPopupForm = CreatePopupForm();
			filterPopupForm.FormClosing += new FormClosingEventHandler(filterPopupFormOnFormClosing);
			PopulateFilterList();
			ShowPopupForm(filterButtonBounds);
		}

		protected virtual List<ColumnFilterItem> GetFilterValues()
		{
			NodesValuesOperation nodesOperation = new NodesValuesOperation(filterShowingColumn);
			OwnerTreeList.NodesIterator.DoOperation(nodesOperation);
			return nodesOperation.Result;
		}

		private static void RemoveColumnDataDuplicates(List<ColumnFilterItem> columnValuesList)
		{
			ColumnFilterItem item = columnValuesList[0];
			int sameIndex = 0;
			int sameCount = 0;
			int i = 1;

			while ( i < columnValuesList.Count )
			{
				if ( columnValuesList[i] == item )
				{
					sameCount++;
					i++;
				} else
				{
					if ( sameCount > 0 )
					{
						columnValuesList.RemoveRange(sameIndex, sameCount);
						i = i - sameCount;
						sameCount = 0;
						sameIndex = i;
						item = columnValuesList[i];
						i++;
					} else
					{
						sameIndex = i;
						item = columnValuesList[i];
						i++;
					}
				}
			}

			if ( sameCount > 0 )
				columnValuesList.RemoveRange(sameIndex, sameCount);
		}

		private static bool IsFilterItemValueEmpty(ColumnFilterItem value)
		{
			return value.IsEmpty;
		}

		protected void PopulateFilterList()
		{
			List<ColumnFilterItem> columnValuesList = GetFilterValues();
			if ( columnValuesList.Count == 0 )
				return;

			columnValuesList.Sort(new FilterItemsComparer());
			columnValuesList.Insert(0, new ColumnFilterCustomItem());
			int blanksItemIndex = 1;
			if ( OwnerTreeList.ColumnFilterConditions.ContainsColumn(filterShowingColumn) )
			{
				columnValuesList.Insert(0, new ColumnFilterAllItem());
				blanksItemIndex++;
			}

			RemoveColumnDataDuplicates(columnValuesList);
			if ( columnValuesList.RemoveAll(IsFilterItemValueEmpty) != 0 )
			{
				columnValuesList.Insert(blanksItemIndex, new ColumnFilterBlanksItem());
				columnValuesList.Insert(blanksItemIndex + 1, new ColumnFilterNonBlanksItem());
			}

			filterPopupForm.filterItemsList.Items.Clear();
			filterPopupForm.filterItemsList.Items.AddRange(columnValuesList.ToArray());
		}

		private void AddSingleFilterCondition(string filterConditionString, ColumnFilterItem filterValue)
		{
			OwnerTreeList.FilteredVisibleNodes.Clear();

			FilterConditionEnum condition;
			ColumnFilterCondition newCondition;

			if ( filterValue is ColumnFilterBlanksItem )
				newCondition = new ColumnFilterCondition(filterShowingColumn.FieldName, FilterConditionEnum.IsBlank, null, null);
			else if ( filterValue is ColumnFilterNonBlanksItem )
				newCondition = new ColumnFilterCondition(filterShowingColumn.FieldName, FilterConditionEnum.IsNotBlank, null, null);
			else
			{
				condition = (FilterConditionEnum)Enum.Parse(typeof(FilterConditionEnum), filterConditionString);
				newCondition = new ColumnFilterCondition(filterShowingColumn.FieldName, condition, filterValue.Value, filterValue.DisplayText);
			}

			for ( int i = 0; i < OwnerTreeList.ColumnFilterConditions.Count; i++ )
				if ( OwnerTreeList.ColumnFilterConditions[i].Column == filterShowingColumn )
				{
					OwnerTreeList.ColumnFilterConditions.RemoveAt(i);
					break;
				}

			if ( !(filterValue is ColumnFilterAllItem) )
				OwnerTreeList.ColumnFilterConditions.Add(newCondition);
		}

		protected virtual CustomFilterForm CreateCustomFilterForm()
		{
			return new CustomFilterForm();
		}

		protected virtual void PopuplateFilterConditions(CustomFilterForm frm)
		{
			frm.filterGroupPanel.Text = filterShowingColumn.GetCaption();
			for ( int i = 0; i < OwnerTreeList.ColumnFilterConditions.Count; i++ )
				if ( OwnerTreeList.ColumnFilterConditions[i].Column == filterShowingColumn )
				{
					frm.cbeFilterConditions.Text = OwnerTreeList.ColumnFilterConditions[i].Condition.ToString();
					frm.teValue.Text = OwnerTreeList.ColumnFilterConditions[i].Value.ToString();
					break;
				}
		}

		protected virtual ColumnFilterItem GetCustomFilterValue(CustomFilterForm frm)
		{
			ColumnFilterItem filterValue;
			try
			{
				filterValue = new ColumnFilterItem(Convert.ChangeType(frm.teValue.EditValue, filterShowingColumn.ColumnType), frm.teValue.Text);
			} catch
			{
				filterValue = new ColumnFilterItem(frm.teValue.Text, frm.teValue.Text);
			}

			return filterValue;
		}

		private void ProcessItemSelection(object sender)
		{
			ColumnFilterItem filterValue = (ColumnFilterItem)((PopupForm)sender).SelectedFilterItem;
			if ( filterValue == null )
				return;

			if ( filterValue is ColumnFilterCustomItem && filterShowingColumn != null )
			{
				CustomFilterForm frm = CreateCustomFilterForm();
				PopuplateFilterConditions(frm);

				if ( frm.ShowDialog() == DialogResult.OK )
				{
					filterValue = GetCustomFilterValue(frm);
					AddSingleFilterCondition(frm.cbeFilterConditions.Text, filterValue);
				}
			} else if ( !filterValue.IsEmpty && filterShowingColumn != null )
				AddSingleFilterCondition("Equals", filterValue);

			filterShowingColumn = null;
		}

		private void filterPopupFormOnFormClosing(object sender, FormClosingEventArgs e)
		{
			ProcessItemSelection(sender);
		}

		public void Dispose()
		{
			ownerTreeList = null;
			filterShowingColumn = null;
		}

		public FilterTreeList OwnerTreeList
		{
			get { return ownerTreeList; }
		}

		public TreeListColumn FilterShowingColumn
		{
			get { return filterShowingColumn; }
		}
	}
}
